modules/tensor_mechanics/src/materials/MultiParameterPlasticityStressUpdate.C

```diff
@@ -0,0 +1,949 @@
+
+  // d(Newton-Raphson params)/d(trial_stress_params) that are used in Jacobian calculations
+  // Here the first _num_sp - 1 Newton-Raphson params are the stress_params
+  // and the _num_sp Newton-Raphson param is gaE
+  std::vector<std::vector<Real>> dnrp_dtrialsp(_num_sp + 1, std::vector<Real>(_num_sp, 0.0));
+
+  // Return-map problem: must apply the following changes in stress_params,
+  // and find the returned stress_params and gaE
+  std::vector<Real> del_stress_params(_num_sp);
+  for (unsigned i = 0; i < _num_sp; ++i)
+    del_stress_params[i] = trial_sp[i] - ok_sp[i];
+
+  Real step_taken = 0.0; // amount of del_stress_params that we've applied and the return-map
+                         // problem has succeeded
+  Real step_size = 1.0;  // potentially can apply del_stress_params in substeps
+  Real gaE_total = 0.0;
+
+  // current values of the yield function, derivatives, etc
+  yieldAndFlow smoothed_q;
+
+  // In the following sub-stepping procedure it is possible that
+  // the last step is an elastic step, and therefore smoothed_q won't
+  // be computed on the last step, so we have to compute it.
+  bool smoothed_q_calculated = false;
+
+  while (step_taken < 1.0 && step_size >= _min_step_size)
+  {
+    if (1.0 - step_taken < step_size)
+      // prevent over-shoots of substepping
+      step_size = 1.0 - step_taken;
+
+    // trial variables in terms of admissible variables
+    for (unsigned i = 0; i < _num_sp; ++i)
+      trial_sp[i] = ok_sp[i] + step_size * del_stress_params[i];
+
+    // initialise variables that are to be found via Newton-Raphson
+    std::vector<Real> current_sp = trial_sp;
+    Real gaE = 0.0;
+
+    // flags indicating failure of Newton-Raphson and line-search
+    int nr_failure = 0;
+    int ls_failure = 0;
+
+    // NR iterations taken in this substep
+    Real step_iter = 0.0;
+
+    // The residual-squared for the line-search
+    Real res2 = 0.0;
+
+    if (step_size < 1.0 && yieldF(trial_sp, ok_intnl) <= _f_tol)
+      // This is an elastic step
+      // The "step_size < 1.0" in above condition is for efficiency: we definitely
+      // know that this is a plastic step if step_size = 1.0
+      smoothed_q_calculated = false;
+    else
+    {
+      // this is a plastic step
+
+      // initialise current_sp, gaE and current_intnl based on the non-smoothed situation
+      std::vector<Real> current_intnl(_num_intnl);
+      initialiseVarsV(trial_sp, ok_intnl, current_sp, gaE, current_intnl);
+      // and find the smoothed yield function, flow potential and derivatives
+      smoothed_q = smoothAllQuantities(current_sp, current_intnl);
+      smoothed_q_calculated = true;
+      calculateRHS(trial_sp, current_sp, gaE, smoothed_q, rhs);
+      res2 = calculateRes2(rhs);
+
+      // Perform a Newton-Raphson with linesearch to get current_sp, gaE, and also smoothed_q
+      while (res2 > _f_tol2 && step_iter < (float)_max_nr_its && nr_failure == 0 && ls_failure == 0)
```

Comment:
(float) _max_nr_its.  That's because step_iter is a float, because _iter[_qp] (a MaterialProperty) is a float because Paraview visualises floats, not ints.  Thanks for pointing this out - i think step_iter should be an unsigned, and i'll do a cast once when i set _iter[_qp].